src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -491,6 +580,124 @@ private int FindEndPosition(ReadOnlySpan<char> input, int i)
             return false;
         }
 
+        /// <summary>Find match end position using the original pattern, end position is known to exist. This version also produces captures.</summary>
+        /// <param name="input">input span</param>
+        /// <param name="i">inclusive start position</param>
+        /// <param name="resultRegisters">out parameter for the final register values, which indicate capture starts and ends</param>
+        /// <returns></returns>
+        private int FindEndPositionCapturing(ReadOnlySpan<char> input, int i, out Registers resultRegisters)
+        {
+            int i_end = input.Length;
+            Registers endRegisters = default(Registers);
+            DfaMatchingState<TSetType>? endState = null;
+
+            // Pick the correct start state based on previous character kind.
+            uint prevCharKind = GetCharKind(input, i - 1);
+            DfaMatchingState<TSetType> state = _initialStates[prevCharKind];
+            // Initialize registers with -1, which means "not seen yet"
+            Registers initialRegisters = new Registers
+            {
+                CaptureStarts = new int[_capsize],
+                CaptureEnds = new int[_capsize],
+            };
+            Array.Fill(initialRegisters.CaptureStarts, -1);
+            Array.Fill(initialRegisters.CaptureEnds, -1);
+
+            if (state.IsNullable(GetCharKind(input, i)))
+            {
+                // Empty match exists because the initial state is accepting.
+                i_end = i - 1;
+                endRegisters.Assign(initialRegisters);
+                endState = state;
+            }
+
+            // Use two maps from state IDs to register values for the current and next set of states.
+            // Note that these maps use insertion order, which is used to maintain priorities between states in a way
+            // that matches the order the backtracking engines visit paths.
+            SparseIntMap<Registers> current = new(), next = new();
```

Comment:
Yes I think that would be good. I think the overhead of this might be relatively slowing down patterns like `(Holmes)`, where the matching itself is very fast.

Reply:
I just went ahead and added the optimization you suggested since it was a small thing. It did reduce the overhead for `(Holmes)` somewhat at least.

There's still one unnecessary register allocation that happens when the registers are stored into `endRegisters`. If the pattern ends exactly when it becomes nullable the copy that currently happens is unnecessary. Fixing it requires massaging the logic a bit, as at the end of the pattern the nullability (triggering assignment of the Registers to `endRegisters`) is noticed one iteration earlier than the fact that no more transitions need to be taken (which happens through all the derivatives resulting in dead end states).